api/src/handlers/websocket/main.go

```diff
@@ -0,0 +1,79 @@
+package websocket
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/gorilla/websocket"
+	"net/http"
+	"encoding/json"
+	"time"
+	"porter/api/src/lib/github"
+	"porter/api/src/handlers/websocket/message"
+	"fmt"
+	"porter/api/src/handlers/websocket/events"
+)
+
+var upgrader = websocket.Upgrader{
+	ReadBufferSize:  1024,
+	WriteBufferSize: 1024,
+	CheckOrigin: func(r *http.Request) bool {
+		return true
+	},
+}
+
+type ReceivedSocketEvent struct {
+	Type string `json:"type"`
+}
+
+func Handler(c *gin.Context) {
+	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
+	if err != nil {
+		return
+	}
+
+	for {
+		_, msg, err := conn.ReadMessage()
+		if err != nil {
+			fmt.Println(err)
+			break
+		}
+§
+		var event ReceivedSocketEvent
+		err = json.Unmarshal(msg, &event)
+		if err != nil {
+			fmt.Println(err)
+			continue
+		}
+
+		go handlerSockerEvent(event, conn)
+	}
+}
+
+func handlerSockerEvent(event ReceivedSocketEvent, conn *websocket.Conn) {
+	switch event.Type {
+		case events.EventGetUsedPorts:
+			messagesChannel := make(chan message.Message)
+			repositoriesChannel := make(chan github.Repository)
+			go usedPortsHandler(repositoriesChannel, messagesChannel)
```

Comment:
It makes sense that `usedPortsHandler` fetches (internally) all of the repos async, but the call for `usedPortsHandler` and then the anonymous func below (which processes the results) should be a "sync", process. you're going to process the results in a sync manner, so it seems like a overhead (of complexity).

IMO you should wait for results from `usedPortsHandler` and then start processing.

Reply:
I see what you mean, a few notes: 
I can rename `usedPortsHandler` to `getRepositories` or something like that, and change its signature to returns an array of repositories.
but what I'm missing here is the `messagesChannel ` which sends the client a log of all the actions